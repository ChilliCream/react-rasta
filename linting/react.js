"use strict";

const ON = true;
const OFF = false;

module.exports = {
  extends: "tslint-react",
  rules: {
    // All
    "jsx-alignment": ON,
    "jsx-boolean-value": ON,
    "jsx-curly-spacing": [ON, "never"],
    "jsx-equals-spacing": [ON, "never"],
    "jsx-key": ON,
    "jsx-no-bind": ON,
    "jsx-no-lambda": OFF,
    "jsx-no-multiline-js": OFF,
    "jsx-no-string-ref": ON,
    "jsx-self-close": ON,
    "jsx-wrap-multiline": ON,
  },
};
