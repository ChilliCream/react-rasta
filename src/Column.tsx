// These imports are needed for declaration files
import * as _React from 'react';
import * as _StyledComponents from 'styled-components';
// -------------------------------------------------------------------
import { InterpolationValue } from 'styled-components';
import { _getBreakpoints } from './BreakpointMap';
import { BreakpointValue, BreakpointValues, PropertyValue } from './BreakpointValue';
import styled from './StyledComponents';
import ThemeProperties, { Theme } from './Theme';
import { _getGutterWidth, _ensureInjectingGlobal, _render } from './Utilities';

export type ColumnSize = 'auto' | 'none' | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type ColumnOrder = 'first' | 'last' | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface ColumnProperties extends ThemeProperties {
  order?: BreakpointValue<ColumnOrder>;
  size?: BreakpointValue<ColumnSize>;
}

const Column = styled.div`
  position: relative;
  width: 100%;
  min-height: 1px;

  ${
    (props: ColumnProperties) => {
      const breakpoints = _getBreakpoints(props!.theme);
      const renderer = {
        order: (value?: PropertyValue) => _renderOrder(value as ColumnOrder),
        size: (value?: PropertyValue) => _renderSize(value as ColumnSize)
      };
      const valueMap = {
        order: props!.order as BreakpointValues<ColumnOrder>,
        size: props!.size as BreakpointValues<ColumnSize>
      };
      const width = _getGutterWidth(props.theme);

      return `
        padding-right: ${width}px;
        padding-left: ${width}px;

        ${_render(valueMap, breakpoints, renderer)}
        ${_renderOrder(props!.order as ColumnOrder)}
        ${_renderSize(props!.size as ColumnSize)}
      `;
    }
  }
`;

export default Column;

function _renderOrder(order?: ColumnOrder | 0 | 13): string {
  if (order != null && typeof order !== 'object') {
    if (order === 'first') {
      return _renderOrder(0);
    }

    if (order === 'last') {
      return _renderOrder(13);
    }

    if (typeof order === 'number') {
      return `
        -webkit-box-ordinal-group: ${order};
        -ms-flex-order: ${order - 1};
        order: ${order - 1};
      `;
    }
  }

  return '';
}

function _renderSize(size?: ColumnSize): string {
  if (size == null || typeof size === 'object' || size < 1 || size > 12) {
    return '';
  }

  if (size === 'none') {
    return `
      -webkit-box-flex: 0;
      -ms-flex: 0 0 auto;
      flex: 0 0 auto;
      width: auto;
      max-width: none;
    `;
  }

  if (size === 'auto') {
    return `
      -ms-flex-preferred-size: 0;
      flex-basis: 0;
      -webkit-box-flex: 1;
      -ms-flex-positive: 1;
      flex-grow: 1;
      max-width: 100%;
    `;
  }

  const percentage = (size / 12 * 100).toFixed(6);

  return `
    -webkit-box-flex: 0;
    -ms-flex: 0 0 ${percentage}%;
    flex: 0 0 ${percentage}%;
    max-width: ${percentage}%;
  `;
}

_ensureInjectingGlobal();