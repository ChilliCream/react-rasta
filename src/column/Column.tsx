import "../utils/bootstrap";
import { ClassAttributes, HTMLAttributes } from "react";
import { StyledComponentClass } from "styled-components";
import ColumnAlignSelf from "./ColumnAlignSelf";
import ColumnOffset from "./ColumnOffset";
import ColumnOrder from "./ColumnOrder";
import ColumnProperties from "./ColumnProperties";
import ColumnSize from "./ColumnSize";
import renderAlignSelf from "./renderAlignSelf";
import renderOffset from "./renderOffset";
import renderOrder from "./renderOrder";
import renderSize from "./renderSize";
import { BreakpointValue, PropertyValue } from "../media";
import { styled, Theme } from "../theme";
import { render } from "../utils";

const Column = styled.div`
  position: relative;
  width: 100%;
  min-height: 1px;

  ${(props: ColumnProperties) => {
    const renderer = {
      alignSelf: (value?: PropertyValue) =>
        renderAlignSelf(value as ColumnAlignSelf),
      offset: (value?: PropertyValue) => renderOffset(value as ColumnOffset),
      order: (value?: PropertyValue) => renderOrder(value as ColumnOrder),
      size: (value?: PropertyValue) => renderSize(value as ColumnSize)
    };
    const valueMap = {
      alignSelf: props!.alignSelf as BreakpointValue<ColumnAlignSelf>,
      offset: props!.offset as BreakpointValue<ColumnOffset>,
      order: props!.order as BreakpointValue<ColumnOrder>,
      size: props!.size as BreakpointValue<ColumnSize>
    };

    return render(valueMap, renderer, props!.theme);
  }};
`;

export default Column;
