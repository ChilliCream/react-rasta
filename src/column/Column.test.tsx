// Module mocks must mocked before importing modules
jest.mock("../utils/bootstrap", () => jest.fn());

import "jest-styled-components";
import { configure, shallow } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import styled from "styled-components";
import Column from "./Column";

configure({ adapter: new Adapter() });

describe("<Column />", () => {
  it("should match the snapshot (no properties set)", () => {
    // act
    const result = shallow(<Column />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (order property set to 5) ", () => {
    // act
    const result = shallow(<Column order={5} />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (order property set to { xs: 'first', sm: 3, md: 'last' }) ", () => {
    // act
    const result = shallow(
      <Column order={{ xs: "first", sm: 3, md: "last" }} />
    );

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (order property set to 5) ", () => {
    // act
    const result = shallow(<Column size={7} />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (size property set to { xs: 'auto', sm: 7, md: 'none' }) ", () => {
    // act
    const result = shallow(<Column size={{ xs: "auto", sm: 7, md: "none" }} />);

    // assert
    expect(result).toMatchSnapshot();
  });

  it("should match the snapshot (order property set to 3 size property is set 9)", () => {
    // act
    const result = shallow(<Column order={3} size={9} />);

    // assert
    expect(result).toMatchSnapshot();
  });
});
