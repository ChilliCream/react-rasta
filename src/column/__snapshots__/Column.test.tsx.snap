// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Column /> should match the snapshot (no properties set) 1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
}

<div
  className="c0"
/>
`;

exports[`<Column /> should match the snapshot (order property set to { xs: 'first', sm: 3, md: 'last' })  1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-box-ordinal-group: 0;
  -ms-flex-order: -1;
  -webkit-order: -1;
  -ms-flex-order: -1;
  order: -1;
}

@media (min-width:576px) {
  .c0 {
    -webkit-box-ordinal-group: 3;
    -ms-flex-order: 2;
    -webkit-order: 2;
    -ms-flex-order: 2;
    order: 2;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-box-ordinal-group: 13;
    -ms-flex-order: 12;
    -webkit-order: 12;
    -ms-flex-order: 12;
    order: 12;
  }
}

<div
  className="c0"
  order={
    Object {
      "md": "last",
      "sm": 3,
      "xs": "first",
    }
  }
/>
`;

exports[`<Column /> should match the snapshot (order property set to 3 size property is set 9) 1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-box-ordinal-group: 3;
  -ms-flex-order: 2;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-box-flex: 0;
  -ms-flex: 0 0 75.000000%;
  -webkit-flex: 0 0 75.000000%;
  -ms-flex: 0 0 75.000000%;
  flex: 0 0 75.000000%;
  max-width: 75.000000%;
}

<div
  className="c0"
  order={3}
  size={9}
/>
`;

exports[`<Column /> should match the snapshot (order property set to 5)  1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-box-ordinal-group: 5;
  -ms-flex-order: 4;
  -webkit-order: 4;
  -ms-flex-order: 4;
  order: 4;
}

<div
  className="c0"
  order={5}
/>
`;

exports[`<Column /> should match the snapshot (order property set to 5)  2`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-box-flex: 0;
  -ms-flex: 0 0 58.333333%;
  -webkit-flex: 0 0 58.333333%;
  -ms-flex: 0 0 58.333333%;
  flex: 0 0 58.333333%;
  max-width: 58.333333%;
}

<div
  className="c0"
  size={7}
/>
`;

exports[`<Column /> should match the snapshot (size property set to { xs: 'auto', sm: 7, md: 'none' })  1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 15px;
  padding-left: 15px;
  -ms-flex-preferred-size: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
}

@media (min-width:576px) {
  .c0 {
    -webkit-box-flex: 0;
    -ms-flex: 0 0 58.333333%;
    -webkit-flex: 0 0 58.333333%;
    -ms-flex: 0 0 58.333333%;
    flex: 0 0 58.333333%;
    max-width: 58.333333%;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-box-flex: 0;
    -ms-flex: 0 0 auto;
    -webkit-flex: 0 0 auto;
    -ms-flex: 0 0 auto;
    flex: 0 0 auto;
    width: auto;
    max-width: none;
  }
}

<div
  className="c0"
  size={
    Object {
      "md": "none",
      "sm": 7,
      "xs": "auto",
    }
  }
/>
`;
